Improve UnaliasSymbolReference for ExchangeNode  and JoinNode

sopel39 opened the pull request:
FYI: @martint 

Peek memory improvements (MBs, no compaction):

```
                       Q5          Q2          Q3         Q7        Q8           Q9        Q10        Q17
Optimized             397         57.9        131        212       480          617       85.5        293
No optimizations      527         85          213        282       699          885       115         397
```

sopel39 commented:
Symbol unaliasing for ExchangeNode canonizes
symbols that are aliased in source nodes.

Plan after optimization:
```
presto:default> explain SELECT c.custkey FROM customer c, orders o WHERE c.custkey = o.custkey AND o.orderdate >= DATE '1994-01-01';
                                                                                  Query Plan
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 - Output[custkey] => [custkey:bigint]
     - RemoteExchange[GATHER] => custkey:bigint
         - Project => [custkey:bigint]
             - InnerJoin[("custkey" = "custkey_0")] => [custkey:bigint, $hashvalue:bigint, custkey_0:bigint, $hashvalue_15:bigint]
                 - RemoteExchange[REPARTITION] => custkey:bigint, $hashvalue:bigint
                     - Project => [custkey:bigint, $hashvalue_14:bigint]
                             $hashvalue_14 := "combine_hash"(BIGINT '0', COALESCE("$operator$hash_code"("custkey"), 0))
                         - TableScan[hive:hive:default:customer, originalConstraint = true] => [custkey:bigint]
                                 LAYOUT: hive
                                 custkey := HiveColumnHandle{clientId=hive, name=custkey, hiveType=bigint, hiveColumnIndex=0, columnType=REGULAR}
                 - RemoteExchange[REPARTITION] => custkey_0:bigint, $hashvalue_15:bigint
                     - Project => [$hashvalue_16:bigint, custkey_0:bigint]
                             $hashvalue_16 := "combine_hash"(BIGINT '0', COALESCE("$operator$hash_code"("custkey_0"), 0))
                         - Filter[("orderdate" >= "$literal$date"(BIGINT '8766'))] => [custkey_0:bigint, orderdate:date]
                             - TableScan[hive:hive:default:orders, originalConstraint = ("orderdate" >= "$literal$date"(BIGINT '8766'))] => [custkey_0:bigint, orderdate:date]
                                     LAYOUT: hive
                                     custkey_0 := HiveColumnHandle{clientId=hive, name=custkey, hiveType=bigint, hiveColumnIndex=1, columnType=REGULAR}
                                     orderdate := HiveColumnHandle{clientId=hive, name=orderdate, hiveType=date, hiveColumnIndex=4, columnType=REGULAR}
```

Plan before optimization:
```
presto:default> explain SELECT c.custkey FROM customer c, orders o WHERE c.custkey = o.custkey AND o.orderdate >= DATE '1994-01-01';
                                                                                    Query Plan
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 - Output[custkey] => [custkey:bigint]
     - RemoteExchange[GATHER] => custkey:bigint
         - Project => [custkey:bigint]
             - InnerJoin[("custkey_8" = "custkey_9")] => [custkey:bigint, custkey_8:bigint, $hashvalue:bigint, custkey_0:bigint, custkey_9:bigint, $hashvalue_16:bigint]
                 - Project => [custkey:bigint, custkey_8:bigint, $hashvalue:bigint]
                     - RemoteExchange[REPARTITION] => custkey:bigint, custkey_8:bigint, $hashvalue:bigint, $hashvalue_14:bigint
                         - Project => [custkey:bigint, $hashvalue_15:bigint]
                                 $hashvalue_15 := "combine_hash"(BIGINT '0', COALESCE("$operator$hash_code"("custkey"), 0))
                             - TableScan[hive:hive:default:customer, originalConstraint = true] => [custkey:bigint]
                                     LAYOUT: hive
                                     custkey := HiveColumnHandle{clientId=hive, name=custkey, hiveType=bigint, hiveColumnIndex=0, columnType=REGULAR}
                 - Project => [custkey_0:bigint, custkey_9:bigint, $hashvalue_16:bigint]
                     - RemoteExchange[REPARTITION] => custkey_0:bigint, custkey_9:bigint, $hashvalue_16:bigint, $hashvalue_17:bigint
                         - Project => [$hashvalue_18:bigint, custkey_0:bigint]
                                 $hashvalue_18 := "combine_hash"(BIGINT '0', COALESCE("$operator$hash_code"("custkey_0"), 0))
                             - Filter[("orderdate" >= "$literal$date"(BIGINT '8766'))] => [custkey_0:bigint, orderdate:date]
                                 - TableScan[hive:hive:default:orders, originalConstraint = ("orderdate" >= "$literal$date"(BIGINT '8766'))] => [custkey_0:bigint, orderdate:date]
                                         LAYOUT: hive
                                         custkey_0 := HiveColumnHandle{clientId=hive, name=custkey, hiveType=bigint, hiveColumnIndex=1, columnType=REGULAR}
                                         orderdate := HiveColumnHandle{clientId=hive, name=orderdate, hiveType=date, hiveColumnIndex=4, columnType=REGULAR}
```

sopel39 commented:
This improves both memory bandwidth usage (no unnecessary are being sent) as well as memory utilization (less columns in join build table).

kokosing commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/UnaliasSymbolReferences.java:
```diff
@@ -261,18 +262,39 @@ public PlanNode visitExchange(ExchangeNode node, RewriteContext<Void> context)
+                        continue;
+                    }
+
+                    addedOutputs.add(canonicalOutput);
```
please try to extract this optimization as separate one. If these two kind of optimization are mixed together it is hard to follow what is happening here.

kokosing commented on presto-main/src/test/java/com/facebook/presto/sql/planner/TestLogicalPlanner.java:
```diff
@@ -237,6 +238,37 @@ public void testCorrelatedScalarAggregationRewriteToLeftOuterJoin()
     }
 
+    @Test
+    public void testUnaliasesJoinRemoteExchangeSymbols()
```
I would name it as `testNoExtraSymbolsInRemoteExchange`

kokosing commented:
Also please check if TPC-H queries which were failing on travis started to pass (exclude them from the `big_query` group).

sopel39 commented:
Addressed comments

kokosing commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/UnaliasSymbolReferences.java:
```diff
@@ -287,6 +292,45 @@ public PlanNode visitExchange(ExchangeNode node, RewriteContext<Void> context)
+            }
+
+            Symbol firstSourceCanonicalInput = canonicalize(node.getInputs().get(0).get(symbolIndex));
+            Set<Symbol> outputAliases = new HashSet<>(inputToOutputs.get(firstSourceCanonicalInput));
```
why do you put output of `firstSourceCanonicalInput` into `inputToOutputs` for each `symbolIndex`?

kokosing commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/UnaliasSymbolReferences.java:
```diff
@@ -287,6 +292,45 @@ public PlanNode visitExchange(ExchangeNode node, RewriteContext<Void> context)
+            Optional<Symbol> outputSymbolForInputSymbols = getExchangeNodeOutputSymbolForInputSymbols(node, symbolIndex, inputToOutputs);
+
+            if (outputSymbolForInputSymbols.isPresent()
+                    && !canonicalOutput.equals(outputSymbolForInputSymbols.get())) {
```
you can shorten the condition in following way:
```
if (outputSymbolForInputSymbols.map(canonicalOutput::equals).orElse(false))
```

kokosing commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/UnaliasSymbolReferences.java:
```diff
@@ -287,6 +292,45 @@ public PlanNode visitExchange(ExchangeNode node, RewriteContext<Void> context)
+            }
+            else {
+                node.getInputs().stream()
+                        .forEach(input -> inputToOutputs.put(input.get(symbolIndex), canonicalOutput));
```
I  think `inputToOutputs` should contain canonical versions of input and output. That case you will also optimize output when inputs are different but their canonical versions are the same.

sopel39 replied:
yes, missed that during refactoring

kokosing commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/UnaliasSymbolReferences.java:
```diff
@@ -261,13 +264,15 @@ public PlanNode visitExchange(ExchangeNode node, RewriteContext<Void> context)
                     .collect(toImmutableList());
 
             List<List<Symbol>> inputs = new ArrayList<>();
+            Multimap<Symbol, Symbol> inputToOutputs = ArrayListMultimap.create();
```
Instead of such multimap I would have a `Map<List<Symbol>, Symbol> exchangeNodeMapping` - a mapping of canonical inputs to canonical output. I think that way `getExchangeNodeOutputSymbolForInputSymbols(` will become much easier, as you will only canonicalize the input and check if it is are in this mappiing. No need for `outputAliases` variable and using `Set#retainAll` method.

sopel39 replied:
good idea

kokosing commented:
fix (restart) the travis

sopel39 commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/UnaliasSymbolReferences.java:
```diff
@@ -261,18 +262,39 @@ public PlanNode visitExchange(ExchangeNode node, RewriteContext<Void> context)
                     .collect(toImmutableList());
 
             List<List<Symbol>> inputs = new ArrayList<>();
+            List<Map<Symbol, Symbol>> inputToOutputs = new LinkedList<>();
```
not related anymore

sopel39 commented:
Addressed comments

kokosing commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/UnaliasSymbolReferences.java:
```diff
@@ -287,6 +289,38 @@ public PlanNode visitExchange(ExchangeNode node, RewriteContext<Void> context)
+
+        private Optional<Symbol> getExchangeNodeOutputSymbolForInputSymbols(ExchangeNode node, int symbolIndex, Map<List<Symbol>, Symbol> inputsToOutputs)
+        {
+            if (node.getInputs().isEmpty()) {
```
is it possible that inputs are empty when you any output?

sopel39 replied:
I didn't see any assertion in ExchangeNode that there has to be at least 1 source.

kokosing commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/UnaliasSymbolReferences.java:
```diff
@@ -287,6 +289,38 @@ public PlanNode visitExchange(ExchangeNode node, RewriteContext<Void> context)
+            return Optional.ofNullable(inputsToOutputs.get(getExchangeNodeInputSymbols(node, symbolIndex)));
+        }
+
+        private List<Symbol> getExchangeNodeInputSymbols(ExchangeNode node, int symbolIndex)
```
rename method to `canonicalizeExchangeNodeInputs`

kokosing commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/UnaliasSymbolReferences.java:
```diff
@@ -261,13 +261,15 @@ public PlanNode visitExchange(ExchangeNode node, RewriteContext<Void> context)
                     .collect(toImmutableList());
 
             List<List<Symbol>> inputs = new ArrayList<>();
+            Map<List<Symbol>, Symbol> inputsToOutputs = new HashMap<>();
```
It is much better now. It is acceptable to me, but I would do it differently ;)

I would do two `ExchangeNode` transformations. First I would create `ExchangeNode` with canonical inputs and outputs (the old `visitExchangeNode`). Then I would remove input duplicates.

martint replied:
What does this map represent? What's the list in the key?

sopel39 commented:
@martint you might want to take a look at memory improvement figures (first comment).

kokosing commented:
@martint Can you please take a look, this improves memory usage up to 40% (see comment about tpch Q3 above).

It achieves that by removing duplicated symbols from ExchangeNode

KBP-TDC commented:
@sopel39 : please fix a typo in one of the  commit message ("cardinaility").

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/UnaliasSymbolReferences.java:
```diff
@@ -261,13 +261,15 @@ public PlanNode visitExchange(ExchangeNode node, RewriteContext<Void> context)
             ImmutableList.Builder<Symbol> outputs = ImmutableList.builder();
             for (int symbolIndex = 0; symbolIndex < node.getOutputSymbols().size(); symbolIndex++) {
-                Symbol canonicalOutput = canonicalize(node.getOutputSymbols().get(symbolIndex));
+                Symbol canonicalOutput = canonicalizeExchangeNodeOutput(node, symbolIndex, inputsToOutputs);
```
This is confusing and hard to follow. This method processes each output, but on each loop, it modifies the inputsToOutputs map that is used by the next loop. It's hard to reason about the side effects.

sopel39 replied:
extracted into separate phase

sopel39 replied:
added comment

sopel39 commented on presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/PlanMatchPattern.java:
```diff
@@ -152,6 +152,11 @@ public PlanMatchPattern withSymbol(String pattern, String alias)
         return with(new SymbolMatcher(pattern, alias));
     }
 
+    public PlanMatchPattern withNumberOfSymbols(int numberOfSymbols)
```
done

sopel39 commented:
Additionally, I have added two more commits at the end that improve unaliasing in join nodes.
Currently plan for: 
```
EXPLAIN (TYPE DISTRIBUTED)
SELECT
   p.partkey
  FROM
   part p,
   supplier s,
   lineitem l,
   partsupp ps
  WHERE
   s.suppkey = l.suppkey AND
   ps.suppkey = l.suppkey AND
   ps.partkey = l.partkey AND
   p.partkey = l.partkey
```

is nondeterministic and contains one unnecessary stage:
```
 Fragment 3 [HASH]                                                                                                                                                                                          
     Output layout: [orderkey, nationkey, quantity, extendedprice, discount, supplycost, $hashvalue_111]                                                                                                    
     Output partitioning: HASH [orderkey]                                                                                                                                                                   
     - Project => [nationkey:bigint, orderkey:bigint, quantity:double, extendedprice:double, discount:double, supplycost:double, $hashvalue_111:bigint]                                                     
             $hashvalue_111 := "combine_hash"(BIGINT '0', COALESCE("$operator$hash_code"("orderkey"), 0))                                                                                                   
         - InnerJoin[("suppkey_4" = "suppkey_8") AND ("partkey" = "partkey_7")] => [suppkey_4:bigint, partkey:bigint, nationkey:bigint, orderkey:bigint, quantity:double, extendedprice:double, discount:dou
             - RemoteSource[4] => [suppkey_4:bigint, partkey:bigint, nationkey:bigint, orderkey:bigint, quantity:double, extendedprice:double, discount:double, $hashvalue_103:bigint]                      
             - RemoteSource[8] => [suppkey_8:bigint, partkey_7:bigint, supplycost:double, $hashvalue_109:bigint]                                                                                            
                                                                                                                                                                                                            
 Fragment 4 [HASH]                                                                                                                                                                                          
     Output layout: [suppkey_4, partkey, nationkey, orderkey, quantity, extendedprice, discount, $hashvalue_108]                                                                                            
     Output partitioning: HASH [suppkey_4, partkey]                                                                                                                                                         
     - Project => [partkey:bigint, suppkey_4:bigint, nationkey:bigint, orderkey:bigint, quantity:double, $hashvalue_108:bigint, extendedprice:double, discount:double]                                      
             $hashvalue_108 := "combine_hash"("combine_hash"(BIGINT '0', COALESCE("$operator$hash_code"("suppkey_4"), 0)), COALESCE("$operator$hash_code"("partkey"), 0))                                   
         - InnerJoin[("suppkey" = "suppkey_4") AND ("partkey" = "partkey_3")] => [suppkey:bigint, partkey:bigint, nationkey:bigint, $hashvalue_104:bigint, suppkey_4:bigint, partkey_3:bigint, orderkey:bigi
             - RemoteSource[5] => [suppkey:bigint, partkey:bigint, nationkey:bigint, $hashvalue_104:bigint]                                                                                                 
             - RemoteSource[7] => [suppkey_4:bigint, partkey_3:bigint, orderkey:bigint, quantity:double, extendedprice:double, discount:double, $hashvalue_106:bigint] 
```
with symbol aliasing those two stages are optimized to:
```
 Fragment 3 [HASH]                                                                                                                                                                                          
     Output layout: [orderkey, nationkey, quantity, extendedprice, discount, supplycost, $hashvalue_122]                                                                                                    
     Output partitioning: HASH [orderkey]                                                                                                                                                                   
     - Project => [nationkey:bigint, orderkey:bigint, quantity:double, extendedprice:double, discount:double, supplycost:double, $hashvalue_122:bigint]                                                     
             $hashvalue_122 := "combine_hash"(BIGINT '0', COALESCE("$operator$hash_code"("orderkey"), 0))                                                                                                   
         - InnerJoin[("suppkey" = "suppkey_8") AND ("partkey" = "partkey_7")] => [partkey:bigint, nationkey:bigint, orderkey:bigint, quantity:double, $hashvalue_119:bigint, extendedprice:double, discount:
             - Project => [partkey:bigint, nationkey:bigint, orderkey:bigint, quantity:double, $hashvalue_119:bigint, extendedprice:double, discount:double, suppkey:bigint]                                
                     $hashvalue_119 := "combine_hash"("combine_hash"(BIGINT '0', COALESCE("$operator$hash_code"("suppkey"), 0)), COALESCE("$operator$hash_code"("partkey"), 0))                             
                 - InnerJoin[("partkey" = "partkey_3") AND ("suppkey" = "suppkey_4")] => [suppkey:bigint, partkey:bigint, nationkey:bigint, $hashvalue_111:bigint, suppkey_4:bigint, partkey_3:bigint, order
                     - Project => [suppkey:bigint, partkey:bigint, nationkey:bigint, $hashvalue_111:bigint]                                                                                                 
                         - RemoteSource[5] => [suppkey:bigint, partkey:bigint, nationkey:bigint, $hashvalue_111:bigint, $hashvalue_112:bigint]                                                              
                     - Project => [suppkey_4:bigint, partkey_3:bigint, orderkey:bigint, quantity:double, extendedprice:double, discount:double, $hashvalue_115:bigint]                                      
                         - RemoteSource[7] => [suppkey_4:bigint, partkey_3:bigint, orderkey:bigint, quantity:double, extendedprice:double, discount:double, $hashvalue_115:bigint, $hashvalue_116:bigint]   
             - RemoteSource[8] => [suppkey_8:bigint, partkey_7:bigint, supplycost:double, $hashvalue_120:bigint]  
```

kokosing commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/UnaliasSymbolReferences.java:
```diff
@@ -161,7 +161,7 @@ public PlanNode visitGroupId(GroupIdNode node, RewriteContext<Void> context)
             List<List<Symbol>> groupingSetsSymbols = node.getGroupingSets().stream()
                     .map(this::canonicalize)
-                    .collect(Collectors.toList());
+                    .collect(toList());
```
use to `toImmutableList()`

sopel39 replied:
done

kokosing commented on presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/SymbolMatcher.java:
```diff
@@ -28,11 +28,20 @@
     {
         this.pattern = Pattern.compile(pattern);
         this.alias = alias;
+        this.exact = false;
```
call `this`

sopel39 replied:
done

kokosing commented on presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/SymbolMatcher.java:
```diff
@@ -47,7 +56,7 @@ public boolean matches(PlanNode node, Session session, Metadata metadata, Expres
         if (symbol != null) {
             expressionAliases.put(alias, symbol.toSymbolReference());
-            return true;
+            return !exact || expressionAliases.get(alias).size() <= 1;
```
s/<=/==

sopel39 replied:
done

kokosing commented on presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/PlanMatchPattern.java:
```diff
@@ -152,6 +152,11 @@ public PlanMatchPattern withSymbol(String pattern, String alias)
         return with(new SymbolMatcher(pattern, alias));
     }
 
+    public PlanMatchPattern withExactSymbol(String pattern, String alias)
```
can you add some javadoc which explains what `Exact` means here

kokosing commented on presto-main/src/test/java/com/facebook/presto/sql/planner/TestLogicalPlanner.java:
```diff
@@ -237,6 +239,51 @@ public void testCorrelatedScalarAggregationRewriteToLeftOuterJoin()
+                        node(ExchangeNode.class,
+                                anyTree(
+                                        node(TableScanNode.class).withExactSymbol("custkey", "C1")
+                                )
```
move it to the line above

sopel39 replied:
done

kokosing commented on presto-main/src/test/java/com/facebook/presto/sql/planner/TestLogicalPlanner.java:
```diff
@@ -237,6 +239,51 @@ public void testCorrelatedScalarAggregationRewriteToLeftOuterJoin()
+    @Test
+    public void testNoExtraSymbolsInJoinRemoteExchange()
+    {
+        assertDistributedPlan("SELECT c.custkey FROM customer c, orders o WHERE c.custkey = o.custkey", anyTree(
```
move each arg in separate line

sopel39 replied:
done

kokosing commented on presto-main/src/test/java/com/facebook/presto/sql/planner/TestLogicalPlanner.java:
```diff
@@ -237,6 +239,51 @@ public void testCorrelatedScalarAggregationRewriteToLeftOuterJoin()
+    public void testDoesNotAliasUnionRemoteExchangeSymbols()
+    {
+        assertDistributedPlan("(SELECT custkey, custkey FROM orders) UNION ALL (SELECT custkey, orderkey FROM orders)", anyTree(
+                node(ExchangeNode.class, anyTree(), anyTree())
```
put `anyTree()` in separate lines

kokosing commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/UnaliasSymbolReferences.java:
```diff
@@ -521,7 +521,14 @@ public PlanNode visitJoin(JoinNode node, RewriteContext<Void> context)
             PlanNode right = context.rewrite(node.getRight());
 
-            return new JoinNode(node.getId(), node.getType(), left, right, canonicalizeJoinCriteria(node.getCriteria()), node.getFilter().map(this::canonicalize), canonicalize(node.getLeftHashSymbol()), canonicalize(node.getRightHashSymbol()));
+            List<JoinNode.EquiJoinClause> canonicalCriteria = canonicalizeJoinCriteria(node.getCriteria());
```
extend commit message. What is optimized and why.

kokosing commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/UnaliasSymbolReferences.java:
```diff
@@ -521,7 +521,14 @@ public PlanNode visitJoin(JoinNode node, RewriteContext<Void> context)
+            Optional<Symbol> canonicalLeftHashSymbol = canonicalize(node.getLeftHashSymbol());
+            Optional<Symbol> canonicalRightHashSymbol = canonicalize(node.getRightHashSymbol());
+
+            canonicalCriteria.forEach(equiJoinClause -> map(equiJoinClause.getRight(), equiJoinClause.getLeft()));
```
add tests for type only coercions as join condition

sopel39 replied:
done

kokosing commented on presto-main/src/test/java/com/facebook/presto/sql/planner/TestLogicalPlanner.java:
```diff
@@ -284,6 +284,55 @@ public void testDoesNotAliasUnionRemoteExchangeSymbols()
+    {
+        assertDistributedPlan("" +
+                        "  SELECT\n" +
+                        "   ps.partkey,\n" +
```
remove '\n' chars

sopel39 commented:
@martint applied comments

sopel39 commented:
Follow up pull request: https://github.com/prestodb/presto/pull/6398

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/UnaliasSymbolReferences.java:
```diff
@@ -285,6 +286,53 @@ public PlanNode visitExchange(ExchangeNode node, RewriteContext<Void> context)
+            // Mapping from list [node.getInput(0).get(symbolIndex), node.getInput(1).get(symbolIndex), ...] to node.getOutputSymbols(symbolIndex).
+            // All symbols are canonical.
+            Map<List<Symbol>, Symbol> inputsToOutputs = new HashMap<>();
+            for (int symbolIndex = 0; symbolIndex < node.getOutputSymbols().size(); symbolIndex++) {
```
Can you explain how this works? I'm having a hard time reverse-engineering the intent of this code.

sopel39 replied:
Added comment: `Map each same list of input symbols [I1, I2, ..., In] to the same output symbol O`

sopel39 commented on presto-main/src/test/java/com/facebook/presto/sql/planner/assertions/PlanMatchPattern.java:
```diff
@@ -152,6 +152,14 @@ public PlanMatchPattern withSymbol(String pattern, String alias)
+    /**
+     * Binds symbol that matches pattern with a given alias. Only one symbol can be bound to a given alias.
+     */
+    public PlanMatchPattern withExactSymbol(String pattern, String alias)
```
This matcher matches symbol by pattern as `withSymbol`. The difference is that `withExactSymbol` ensures that there is only one symbol reference for a given pattern. This ensures that the alias relates to exactly same symbol in plan (e.g: not symbol aliased by projection, e.g: `custkey => custkey_6`).

sopel39 commented on presto-main/src/main/java/com/facebook/presto/testing/LocalQueryRunner.java:
```diff
@@ -646,14 +646,19 @@ public Plan createPlan(Session session, @Language("SQL") String sql)
         FeaturesConfig featuresConfig = new FeaturesConfig()
                 .setDistributedIndexJoinsEnabled(false)
                 .setOptimizeHashGeneration(true);
-        PlanOptimizers planOptimizers = new PlanOptimizers(metadata, sqlParser, featuresConfig, true);
```
`LocalQueryRunner` can also create plan for a given query. I added option for the plan generation to create plan with Exchanges to test distributed plans. Such distributed plans are not executed by `LocalQueryRunner`

sopel39 commented:
addressed comments, incorporated `Make PredicatePushdown optimizer not create unnecessary symbols` https://github.com/prestodb/presto/pull/6451/commits/f96a39a69b195ea76d7b2a10efc046b9b26a4628 into this PR

martint commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/UnaliasSymbolReferences.java:
```diff
@@ -299,6 +302,61 @@ public PlanNode visitExchange(ExchangeNode node, RewriteContext<Void> context)
+
+        private Optional<Symbol> exchangeNodeOutputSymbolForInputSymbols(ExchangeNode node, List<Symbol> canonicalInputs, Map<List<Symbol>, Symbol> inputsToOutputs)
+        {
+            if (node.getInputs().isEmpty() || !inputsToOutputs.containsKey(canonicalInputs)) {
```
When would `node.getInputs()` be empty?

sopel39 replied:
There is no assertion in `ExchangeNode` that is should be non-empty, so I decided to be on safe side.

ebd2 commented on presto-main/src/test/java/com/facebook/presto/sql/planner/TestLogicalPlanner.java:
```diff
@@ -357,29 +359,148 @@ public void testQuantifiedComparisonNotEqualAny()
         );
     }
 
+    @Test
```
Can we move these into a new test class? TestLogicalPlanner is getting big as it stands and this group of tests seems to share a common them.

sopel39 replied:
Sure. As @martint said, he will merge the changes to optimizers first and then I will refactor those tests after your work is merged.

ebd2 replied:
Thanks!

martint commented:
I made a couple of adjustments to "Improve symbol unaliasing for ExchangeNode" and merged the optimizers

sopel39 commented:
Rebased and left test commits

martint commented:
There are some merge conflicts due to changes in the plan matching framework. Can you rebase again? (sorry 😢 )

sopel39 commented:
I am waiting for `strictOutputSymbolsMatcher` by @ebd2, which is required by tests in this PR. When it's done, I will rebase on top of it.

ebd2 commented:
@sopel39: https://github.com/prestodb/presto/pull/6846 was merged. You should be good to rebase.

sopel39 commented:
If that is not a priority, I will close it. Those tests test quite complex distributed plan and I think they are fragile (can be broken by other rules)
